src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -2335,20 +2340,94 @@ public sealed override void Filter(
                     IValueSet? whenFalseValues,
                     out Tests whenTrue,
                     out Tests whenFalse,
-                    ref bool foundExplicitNullTest)
+                    ref bool foundExplicitNullTest,
+                    BindingDiagnosticBag diagnostics)
                 {
                     var trueBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
                     var falseBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
-                    foreach (var other in RemainingTests)
+                    for (int i = 0; i < RemainingTests.Length; i++)
                     {
-                        other.Filter(builder, test, state, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest);
+                        var other = RemainingTests[i];
+                        other.Filter(builder, test, state, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest, diagnostics);
                         trueBuilder.Add(oneTrue);
                         falseBuilder.Add(oneFalse);
+
+                        if (i > 0
+                            && oneTrue is False
+                            && this is OrSequence { RemainingTests: [Not { Negated: var negated }, ..] }
+                            && isRecognizedPartOfNegated(test, negated))
+                        {
+                            // We report a warning based on heuristic to catch a common mistake (`not ... or <redundant>` pattern)
+                            diagnostics.Add(ErrorCode.WRN_RedundantPattern, getSyntax(other));
+                        }
                     }
 
                     whenTrue = Update(trueBuilder);
                     whenFalse = Update(falseBuilder);
+
+                    return;
+
+                    // Check that the given test, which causes the `or` condition to become redundant,
+                    // is indeed the test being negated by the `not`
+                    static bool isRecognizedPartOfNegated(BoundDagTest test, Tests negated)
+                    {
+                        // If we have something like `not A or ... or B` where the A test being true implies that the B test is false,
+                        // so the B test could only be true when the `not A` test is true,
+                        // then the user probably made a mistake having the `or B` condition
+                        // Also handles `not "literal" or ...`
+                        if (negated is AndSequence { RemainingTests: [One { Test: BoundDagNonNullTest nonNullTest }, One { Test: BoundDagTypeTest typeTest }, ..] }
```

Comment:
It feels unfortunate that we do a "normalization" of the tests at this phase, as it feels like we are already having to work backwards to infer what the user pattern actually looked like.

Reply:
Yes, that's been the biggest challenge for this PR: `CheckConsistentDecision` operates on lowered nodes (`BoundDagTest`) and is invoked during the incremental construction of the DAG. There may be a way to implement a similar method to operate on `BoundPattern` from initial binding instead. It could be simplified somewhat (we don't need to return 4 flags, just 1)